pulp_smash/config.py

```diff
@@ -279,72 +178,151 @@ def read(self, section=None, xdg_config_file=None, xdg_config_dir=None):
-                    'at "{2}" appears to be following the old standard; '
-                    'consider renaming section "{1}" to "{0}".'
-                    .format(self._section, 'default', path)
+                    'Pulp Smash configuration file should follow the systems '
+                    'with roles format. But the configuration file at "{0}" '
+                    'appears to be following the old configuration file '
+                    'format. Consider updating the configuration file to the '
+                    'new system with roles format, run python3 -m pulp_smash '
+                    'for more information.'
+                    .format(path)
                 ),
                 DeprecationWarning
             )
-        elif section is None:
-            section = self._section
-        try:
-            config_section = config_file[section]
-        except KeyError:
-            raise exceptions.ConfigFileSectionNotFoundError(
-                'The Pulp Smash configuration file at {} has no section '
-                'entitled {}.'.format(path, section)
-            )
+            parsed_url = urlparse(pulp.get('base_url'))
+            systems.append(PulpSystem(
+                parsed_url.hostname or '',
+                {
+                    'amqp broker': {'service': 'qpidd'},
+                    'api': {
+                        'scheme': parsed_url.scheme,
+                        'verify': pulp.get('verify'),
+                    },
+                    'mongod': {},
+                    'pulp cli': {},
+                    'pulp celerybeat': {},
+                    'pulp resource manager': {},
+                    'pulp workers': {},
+                    'shell': {
+                        'transport': pulp.get('cli_transport'),
+                    },
+                    'squid': {},
+                }
+            ))
+        else:
+            for system in config_file.get('systems', []):
+                systems.append(PulpSystem(**system))
+        return PulpSmashConfig(pulp_auth, pulp_version, systems)
+
+    def get_systems(self, role):
+        """Return a list of systems fulfilling the given role.
 
-        # Instantiate a config and populate it with values from the settings
-        # file. We tell the config object which file it has been populated from
-        # so calls to its `save` method and other methods hit the same file.
-        cfg = type(self)(**config_section)
-        # pylint:disable=protected-access
-        cfg._section = section
-        cfg._xdg_config_file = xdg_config_file
-        cfg._xdg_config_dir = xdg_config_dir
-        return cfg
-
-    def get_requests_kwargs(self):
+        :param role: The role to filter the available systems, see
+            `pulp_smash.config.ROLES` for more information.
+        """
+        return [
+            system for system in self.systems
+            if role in ROLES and role in system.roles
+        ]
```

Comment:
I see the `if role in ROLES` here, and wonder if this would be better:

```python
if role not in ROLES:
    raise ValueError('The given role, {}, is not recognized. Valid roles are: {}'.format(role, ROLES))
return [system for system in self.systems if role in system.roles]
```

The current code will just mysteriously ignore certain `role` arguments, and the user will never be told why.

Alternately, we could just raise a warning if the given `role` isn't recognized, and drop the `if role in ROLES` check.